docs/readmes/lte/GTPU_Extension_Header.md

```diff
@@ -0,0 +1,149 @@
+
+Ryu is a Python library that provides an API wrapper for programming OVS.
+
+There are two parts to taking care of QFI functionality support:
+
+- Nicira Extension Actions Structures (NXAction)
+- Nicira Extended Match Structures (OFPMatch)
+
+### OVS Openflow
+
+Using the OpenFlow protocol, the controller can add, update, and delete flow entries in flow tables, both reactively (in response to packets) and proactively. Each flow table in the switch contains a set of flow entries; each flow entry consists of match fields, counters, and a set of instructions to apply to matching packets.
+
+### OVS Kernel
+
+For Egress flow, one needs to build the extension header and set the QFI value. It needs to be passed down to kernel datapath for pushing it to GTP-U packets.
+For Ingress flow, extract the QFI value from skb of GTP header and set to the tunnel key.
+
+### OVS Command Line
+
+Using below ovs-ofctl commands, will add/del the uplink and downlink flows with QFI parameter.
+
+Uplink add flow:
+
+```sudo ovs-ofctl add-flow gtp_br0 "table=0, priority=100,tun_id=0x7,qfi=6,in_port=gtp0 actions=set_field:02:00:00:00:00:01->eth_src,set_field:ff:ff:ff:ff:ff:ff->eth_dst,set_field:0->reg9,set_field:0x181c9->metadata,resubmit(,1)"```
+
+Downlink add flow:
+
+```sudo ovs-ofctl add-flow gtp_br0 "table=0, priority=65503,ip,in_port=4294967294,nw_dst=192.168.128.11 actions=set_field:0x2710->tun_id,set_field:192.168.60.16->tun_dst,set_field:0x8000->reg8,set_field:0x6->qfi,set_field:oam->tun_flags,set_field:0x181c9->metadata,resubmit(,1)"```
+
+Uplink del flow:
+
+```sudo ovs-ofctl del-flows gtp_br0 "table=0, tun_id=0x7,qfi=6,in_port=gtp0"```
+
+Downlink del flow:
+
+```sudo ovs-ofctl del-flows gtp_br0 "table=0, ip,in_port=4294967294,nw_dst=192.168.128.11"```
+
+## High Level Call Flow
+
+Call flow for QFI set in OVS.
+
+```mermaid
+sequenceDiagram
+    CPE->>MME(AMF): PDU Session Establishment Request
+    MME(AMF)->>SESSIOND: SetSMFContext
+    SESSIOND->>PIPELINED: SetSMFSession(GTP)
+    PIPELINED->>RYU_CONTROLLER: Set QFI value in match and action criteria
+    RYU_CONTROLLER->>OPENFLOW: NXM_NX_QFI
+    OPENFLOW->>OVS_DATAPATH: MFF_QFI and set tunnel QFI
+    CPE->>OVS_DATAPATH: GTP Traffic for uplink/downlink with QFI value
+```
+
+## Common Issues and Troubleshooting
+
+1. GTP kernel module is included as part of OVS module. So no need to insert gtp.ko.
+
+2. To set up a development environment, run the following command on the magma-dev VM.
+   ```sudo bash ~/magma/third_party/gtp_ovs/ovs-gtp-patches/2.15/dev.sh setup```
+
+3. To run OVS GTP tests on OVS kernel datapath:
+   ```sudo bash ~/magma/third_party/gtp_ovs/ovs-gtp-patches/2.15/dev.sh build_test```
+
+4. OVS debug logging can be dynamically enabled by ```sudo ovs-appctl vlog/set dbg```.
+   For a specific module,
+   ```sudo ovs-appctl vlog/set netdev dbg```
+   ```sudo ovs-appctl vlog/set ofproto dbg```
+   ```sudo ovs-appctl vlog/set vswitchd dbg```
+   ```sudo ovs-appctl vlog/set dpif dbg```

```

Comment:
Same comment as above regarding it being put in one block so it's over multiple lines.

Reply:
fixed it